src/hypothesis/extra/pandas/impl.py

```diff
@@ -0,0 +1,382 @@
+# coding=utf-8
+#
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis-python
+#
+# Most of this work is copyright (C) 2013-2017 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at http://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from __future__ import division, print_function, absolute_import
+
+from collections import Counter, Iterable, OrderedDict
+
+import numpy as np
+
+import pandas
+import hypothesis.strategies as st
+import hypothesis.extra.numpy as npst
+from hypothesis.errors import InvalidArgument
+from hypothesis.internal.compat import hrange, integer_types
+
+
+def is_sequence(c):
+    return hasattr(c, '__len__') and hasattr(c, '__getitem__')
+
+
+def is_category_dtype(dtype):
+    # We need to explicitly check that this is not a dtype because a
+    # numpy dtype compared to a string will try to convert it to an
+    # numpy dtype and error if it can't.
+    if isinstance(dtype, np.dtype):
+        return False
+    return dtype == 'category'
+
+
+def build_index(draw, index, min_size, max_size):
+    st.check_valid_interval(
+        min_size, max_size, 'min_size', 'max_size'
+    )
+
+    index_from_strategy = False
+
+    if isinstance(index, st.SearchStrategy):
+        index = draw(index)
+        index_from_strategy = True
+
+    if index is not None:
+        if not is_sequence(index):
+            raise InvalidArgument(
+                '%s was %r of type %s, but expected a sequence' % (
+                    'Result of index strategy' if index_from_strategy
+                    else 'index argument',
+                    index, type(index).__name__
+                )
+            )
+
+        if max_size is None:
+            max_size = len(index)
+        elif index_from_strategy:
+            max_size = min(max_size, len(index))
+        else:
+            raise InvalidArgument((
+                'Provided index %r only has %d elements, which is not '
+                'enough for the provided max_size of %d. Either increase '
+                'the number of elements in the index or reduce or remove '
+                'the max_size argument.') % (
+                    index, len(index), max_size
+            ))
+
+    if max_size is None:
+        max_size = max(10, min_size + 1)
```

Comment:
No especially good reason. It's double our average list length, so it means the typical size range is the same as that for lists. But mostly it's small enough to get reasonable performance by default, large enough to explore a reasonable range of behaviour.